### PR TITLE
Feature/add user interests

### PR DESCRIPTION
**Pull request by Snowflake8**
Add the possibility of the user editing his interests
## Pull reviewers stats
Stats for the last 180 days:
|                                                                                                                                                                 | User            | Total reviews | Median time to review                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | Total comments |
| --------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------- | ------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ | -------------- |
| <a href="https://github.com/so7fie"><img src="https://avatars.githubusercontent.com/u/10992808?v=4" width="20"></a>                                             | so7fie          | 29            | [13h 38m](https://app.flowwer.dev/charts/review-time/~(u~(i~'10992808~n~'so7fie)~p~180~r~(~(d~'qtgafe~t~'rgf)~(d~'qpuy0b~t~'1v)~(d~'qq0p5z~t~'5hby)~(d~'qq0t0v~t~'5m7q)~(d~'qq0t2c~t~'5m97)~(d~'qqew7j~t~'2pkp)~(d~'qqf92s~t~'32fy)~(d~'qqjbho~t~'1ux)~(d~'qrnshx~t~'193)~(d~'qrpf7u~t~'11ut)~(d~'qrr5wv~t~'1e1c)~(d~'qq7qp0~t~'hf)~(d~'qrm6v4~t~'6ai)~(d~'qqkh75~t~'dr0)~(d~'qqrud7~t~'7qx2)~(d~'qr8kh3~t~'1kfr)~(d~'qrm9jc~t~'1npk)~(d~'qrmbez~t~'1ls)~(d~'qqy5l4~t~'5im1)~(d~'qs131e~t~'ghj)~(d~'qs1322~t~'gi7)~(d~'qs6mfb~t~'7b7e)~(d~'qs7tjn~t~'1500)~(d~'qs7udr~t~'12hc)~(d~'qs83lz~t~'8nr)~(d~'qs2ves~t~'18k)~(d~'qsi0yg~t~'676f)~(d~'qt06zc~t~'1ytf)~(d~'qt07rb~t~'4hjd)~(d~'qsrail~t~'muw)~(d~'qsmoye~t~'1g9l)~(d~'qson7e~t~'6g7)~(d~'qson80~t~'6gt)~(d~'qt1j6b~t~'wjg)~(d~'qt285i~t~'3fw)~(d~'qsi275~t~'1yxs)~(d~'qsxwcl~t~'ht38)~(d~'qt1jzy~t~'1d2y)~(d~'qsra8x~t~'hn6)~(d~'qrxija~t~'1azf)~(d~'qrxjn8~t~'1c3d)~(d~'qryfmg~t~'k5v)~(d~'qrygrv~t~'lba)~(d~'qrygwl~t~'lg0)~(d~'qsomvk~t~'53m))))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | **60**         |
| <a href="https://github.com/Snowflake8"><img src="https://avatars.githubusercontent.com/u/44499990?v=4" width="20"></a>                                         | Snowflake8      | 27            | [7h 23m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499990~n~'Snowflake8)~p~180~r~(~(d~'qshau1~t~'19dg)~(d~'qtbd3y~t~'43)~(d~'qpt3s9~t~'2zf8)~(d~'qpx29n~t~'1vgi)~(d~'qq0q7g~t~'5jeb)~(d~'qq6k9y~t~'8hk)~(d~'qq6kh9~t~'8ov)~(d~'qq6r5v~t~'6b)~(d~'qqdm38~t~'1fge)~(d~'qq6ncv~t~'8609)~(d~'qq8el9~t~'y2h)~(d~'qqjh1g~t~'gox)~(d~'qqjh87~t~'7lg)~(d~'qqjlvf~t~'b)~(d~'qro2s6~t~'666)~(d~'qrm4bb~t~'1ljt)~(d~'qrnhv0~t~'7)~(d~'qrkbo5~t~'88t)~(d~'qrm4eh~t~'1rq)~(d~'qrz0il~t~'tz)~(d~'qqsnpq~t~'qn)~(d~'qqxzb3~t~'ikt)~(d~'qrz0hs~t~'27o6)~(d~'qrzb49~t~'3lg2)~(d~'qsatcu~t~'rli)~(d~'qs396k~t~'f0c)~(d~'qsasiz~t~'5yol)~(d~'qsbj9i~t~'6pf4)~(d~'qt06zp~t~'4grr)~(d~'qt0vi6~t~'55a8)~(d~'qt23cw~t~'khx)~(d~'qsybaj~t~'9bt1)~(d~'qsp2w6~t~'d)~(d~'qtt97x~t~'c0)~(d~'qtgthr~t~'31k2))))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                      | 55             |
| <a href="https://github.com/Huntsekkern"><img src="https://avatars.githubusercontent.com/u/44240415?v=4" width="20"></a>                                        | Huntsekkern     | 31            | [**5h 2m**](https://app.flowwer.dev/charts/review-time/~(u~(i~'44240415~n~'Huntsekkern)~p~180~r~(~(d~'qt3udr~t~'gj)~(d~'qtff6r~t~'m)~(d~'qtfe60~t~'hck)~(d~'qtipnq~t~'2avm)~(d~'qsn1ou~t~'3hvc)~(d~'qpumtn~t~'1745)~(d~'qpt0zy~t~'2wmx)~(d~'qpt88n~t~'33vm)~(d~'qptd1d~t~'29r)~(d~'qptd3y~t~'2cc)~(d~'qpt4tp~t~'2b1)~(d~'qpuo7b~t~'mk)~(d~'qq6uip~t~'3j5)~(d~'qq8faa~t~'qv)~(d~'qq8fiz~t~'bv)~(d~'qqjij5~t~'i6m)~(d~'qrs0pv~t~'3yyp)~(d~'qrvgff~t~'1lvl)~(d~'qrihqg~t~'1wf3)~(d~'qrm9wa~t~'17x)~(d~'qrma4k~t~'1g7)~(d~'qrmajv~t~'1vi)~(d~'qrni3f~t~'15)~(d~'qqkomy~t~'z8i)~(d~'qqkozj~t~'zl3)~(d~'qqkp4h~t~'zq1)~(d~'qqvo8r~t~'14d)~(d~'qrmb5j~t~'1cc)~(d~'qrwzfg~t~'6lu)~(d~'qsbchv~t~'hml)~(d~'qrmbxz~t~'1bu)~(d~'qrmc41~t~'1hw)~(d~'qsbhva~t~'4tbn)~(d~'qsbjxk~t~'cx)~(d~'qs85j9~t~'al1)~(d~'qscdmi~t~'688)~(d~'qsdcs5~t~'lp3)~(d~'qs858h~t~'3be3)~(d~'qsvu6v~t~'x8x)~(d~'qt24yw~t~'1xeb)~(d~'qsvqq1~t~'532c)~(d~'qsvqgj~t~'6qz1)~(d~'qt1jgc~t~'wth)~(d~'qt2icu~t~'816)~(d~'qsvrbx~t~'y7p)~(d~'qryy4h~t~'12nw)~(d~'qryx5p~t~'ah5)~(d~'qttbyh~t~'114))))                                                                                                                                                                                                                                                                                                                                                                                                                                                         | 54             |
| <a href="https://github.com/DGoedtkindtEPFL"><img src="https://avatars.githubusercontent.com/u/61052800?v=4" width="20"></a>                                    | DGoedtkindtEPFL | **40**        | [15h 26m](https://app.flowwer.dev/charts/review-time/~(u~(i~'61052800~n~'DGoedtkindtEPFL)~p~180~r~(~(d~'qt9ofl~t~'5uid)~(d~'qtewwa~t~'23p)~(d~'qtomtk~t~'69)~(d~'qtio32~t~'1q)~(d~'qsmngd~t~'33mv)~(d~'qtgalr~t~'rms)~(d~'qtefvx~t~'u8z)~(d~'qtefzp~t~'ucr)~(d~'qtegm0~t~'uz2)~(d~'qpuy16~t~'2q)~(d~'qpuoy1~t~'co)~(d~'qpup0b~t~'19at)~(d~'qplma9~t~'1e2x)~(d~'qpuugt~t~'am9h)~(d~'qq6s0w~t~'8gm)~(d~'qq6rfm~t~'fn8)~(d~'qqjjxz~t~'9s4m)~(d~'qq9yj2~t~'1i5c)~(d~'qqjmgc~t~'dy)~(d~'qqrtyv~t~'xlr)~(d~'qrnxgy~t~'1wd)~(d~'qrsrti~t~'4q2c)~(d~'qqxoq1~t~'21ln)~(d~'qqioj1~t~'2x8a)~(d~'qqionj~t~'2xcs)~(d~'qqioou~t~'2xe3)~(d~'qqipcq~t~'2y1z)~(d~'qqru2n~t~'7qmi)~(d~'qrm4pz~t~'1iw7)~(d~'qrm4za~t~'1j5i)~(d~'qrm5eq~t~'1jky)~(d~'qrmfv4~t~'1u1c)~(d~'qrmhyq~t~'1cp)~(d~'qs0a3d~t~'xnt)~(d~'qs2og9~t~'21we)~(d~'qrxk77~t~'25m)~(d~'qryk8c~t~'58f)~(d~'qs2oux~t~'1k7w)~(d~'qrxjta~t~'qzo)~(d~'qsa5hg~t~'1a0u)~(d~'qsa8mh~t~'1j)~(d~'qsa8rj~t~'6l)~(d~'qsa8tq~t~'8s)~(d~'qsam7t~t~'dmv)~(d~'qscac2~t~'442)~(d~'qsutl8~t~'2l0)~(d~'qt1duk~t~'169z)~(d~'qsp2e9~t~'8z4w)~(d~'qt1ec8~t~'17f8)~(d~'qryv6t~t~'8i9)~(d~'qsogu4~t~'17yr)~(d~'qsovo9~t~'3qz)~(d~'qs0amr~t~'pgk)~(d~'qs0aqu~t~'pkn)~(d~'qs44fc~t~'19vi)~(d~'qtnp2h~t~'35a1)~(d~'qtq1kd~t~'5hrx)~(d~'qtrdjr~t~'10o)~(d~'qtrdky~t~'11v)~(d~'qtrdlj~t~'12g)~(d~'qtrx3k~t~'fpk)~(d~'qtq0xr~t~'1qnr)~(d~'qttczq~t~'1j67)~(d~'qtq227~t~'b0ic)~(d~'qttags~t~'q)~(d~'qtnwp1~t~'a4rc)~(d~'qtnwr2~t~'a4td)~(d~'qtnybu~t~'a6e5)~(d~'qtoc0u~t~'ak35)~(d~'qtrxr1~t~'e5tc)))) | 47             |
| <a href="https://github.com/RaulinN"><img src="https://avatars.githubusercontent.com/u/44499915?u=c029696ff2de0902f905fbf35188359e94e08d88&v=4" width="20"></a> | RaulinN         | 33            | [7h 28m](https://app.flowwer.dev/charts/review-time/~(u~(i~'44499915~n~'RaulinN)~p~180~r~(~(d~'qtfq0z~t~'m)~(d~'qtdqyu~t~'x1o)~(d~'qtnoui~t~'2vgc)~(d~'qtnp1h~t~'50f2)~(d~'qtdpp8~t~'4cnp)~(d~'qtdr1i~t~'18om)~(d~'qt3m50~t~'11g)~(d~'qtdq1w~t~'2d21)~(d~'qptcqf~t~'1yt)~(d~'qpq5rb~t~'5xjz)~(d~'qq73ig~t~'jy6)~(d~'qq7vmj~t~'1t5)~(d~'qpuwak~t~'8pt)~(d~'qq73am~t~'28)~(d~'qqhz4w~t~'87bj)~(d~'qq73x3~t~'8mkh)~(d~'qq821x~t~'lj5)~(d~'qqr6wy~t~'aju)~(d~'qrs0cu~t~'3ylo)~(d~'qrryuq~t~'428q)~(d~'qrrz93~t~'42n3)~(d~'qrrzpp~t~'3mco)~(d~'qrsjsf~t~'7tp)~(d~'qrni37~t~'x)~(d~'qq7qbn~t~'1bz)~(d~'qqj0lo~t~'u5)~(d~'qqjmtu~t~'n2b)~(d~'qrm5d5~t~'2qe)~(d~'qreyev~t~'f)~(d~'qsb1c8~t~'sra)~(d~'qsb2b4~t~'7fu)~(d~'qsb1ye~t~'1072)~(d~'qsciv8~t~'j)~(d~'qsb2ae~t~'2k2)~(d~'qsc5gk~t~'az)~(d~'qt3hia~t~'1cpm)~(d~'qtru7o~t~'2x0i)~(d~'qtrtsa~t~'cea)~(d~'qtrtvf~t~'3jlf)~(d~'qttcx7~t~'1zu))))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | 45             |
| <a href="https://github.com/Laniakea1999"><img src="https://avatars.githubusercontent.com/u/43450242?v=4" width="20"></a>                                       | Laniakea1999    | 22            | [12h 8m](https://app.flowwer.dev/charts/review-time/~(u~(i~'43450242~n~'Laniakea1999)~p~180~r~(~(d~'qtd4xz~t~'b0t)~(d~'qtd42x~t~'3r1e)~(d~'qtd55e~t~'msi)~(d~'qtd61v~t~'9by1)~(d~'qtd3u7~t~'1quc)~(d~'qpupsq~t~'d4)~(d~'qq7oa5~t~'1n)~(d~'qqeyk1~t~'2rx7)~(d~'qrmbac~t~'1h5)~(d~'qs2ml9~t~'201e)~(d~'qrzab0~t~'vb3)~(d~'qrx80e~t~'f6s)~(d~'qsc6gz~t~'coy)~(d~'qsc6o4~t~'7ctq)~(d~'qsl5yv~t~'52pi)~(d~'qrx7lz~t~'1024)~(d~'qsoj15~t~'1a5s)~(d~'qsopgk~t~'1kf)~(d~'qtruu7~t~'2xn1)~(d~'qtrw2v~t~'cuj0)~(d~'qtt9rp~t~'m)~(d~'qttbsi~t~'1cg))))                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        | 15             |

### CODE REVIEW

**Comment by Snowflake8** [comment]
Pas sure de pourquoi ça passe vu que ça marche en local? Je vais peut-être attendre sur @Huntsekkern

**Comment by Huntsekkern** [comment]
> Pas sure de pourquoi ça passe vu que ça marche en local? Je vais peut-être attendre sur @Huntsekkern

> No logs available!
is not something I have much of an input on ^^ I can try locally but you said it worked on your side, and also verify how you did the merge, but apart from that, all my fixes have been pushed onto main

File: app/src/main/java/com/github/polybooks/activities/EditUserInterestsActivity.kt [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    override fun onFilterButtonClick() {
+        startActivity(Intent(this, UserProfileActivity::class.java))
+    }
+
+    override fun adapter(list: List<Interest>): RecyclerView.Adapter<*> {
+        return adapter
+    }
+
+    override fun getQuery(): Query<Interest> {
```
Comment (DGoedtkindtEPFL) [inline app/src/main/java/com/github/polybooks/activities/EditUserInterestsActivity.kt:35]: Ce que j'ai fait dans la PR #227 c'est d'avoir une méthode `getElement()` au lieu d'une méthode `getQuery()` ça évite ce code awkward où on crée une query comme ça.

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+import com.github.polybooks.R
+import com.github.polybooks.core.Interest
+import com.github.polybooks.database.Database
+import com.github.polybooks.utils.StringsManip.getName
+import com.github.polybooks.utils.fireBaseUsertoUser
+import com.google.firebase.auth.FirebaseAuth
+
+class InterestAdapter: RecyclerView.Adapter<InterestAdapter.InterestHolder>(){
+
+    private val user = fireBaseUsertoUser(FirebaseAuth.getInstance().currentUser)
```
Comment (DGoedtkindtEPFL) [inline app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:18]: Ce n'est pas une bonne chose d'appeler `FirebaseAuth.getInstance()` parce que ça ne s'adapte pas dans les tests pour utiliser un mock de Firebase. Il faudrait faire une fonction dans `FirebaseProvider` pour obtenir une instance de FirebaseAuth.

File: app/src/androidTest/java/com/github/polybooks/activities/EditUserInterestsTest.kt [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+import org.junit.Test
+
+class EditUserInterestsTest {
+    val mockFields: List<Field> = listOf(
+        Field("Biology"),
+        Field("Computer Science"),
+        Field("Architecture")
+    )
+
+    val mockCourses: List<Course> = listOf(
```
Comment (DGoedtkindtEPFL) [inline app/src/androidTest/java/com/github/polybooks/activities/EditUserInterestsTest.kt:25]: Est-ce que tu pourrais pas utiliser `Database.interestDatabase` ?

Comment (Huntsekkern) [reply app/src/androidTest/java/com/github/polybooks/activities/EditUserInterestsTest.kt:25]: Database.interestDatabase retourne DummyInterestDatabase car FBInterestDB n'est pas preloadée dans les mocks.

Après, tu voulais peut-être juste dire que ne pas ré-instancier ces 3 listes et les accéder avec `Database.interestDatabase.mockCourses` auquel, je suis d'accord avec toi

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    private var userInterests = Database.interestDatabase.getUserInterests(user)
+        .thenApply { triple -> triple.first + triple.second + triple.third }.get().toMutableSet()
+    private val interests : List<Interest> = Database.interestDatabase.listAllInterests()
+        .thenApply { triple -> triple.first + triple.second + triple.third }.get()
```
Comment (Huntsekkern) [inline app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:22]: Est-ce qu'on veut vraiment une seule liste d'intérêt ici et pas plutôt une liste par type d'intérêt?

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 2 above)
Comment (Huntsekkern) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:18]: 100% d'accord, après, je pense que je vais très bientôt merger sa branche dans la mienne, ça sera plus simple pour la suite, donc je peux m'en occuper si besoin.

Aussi au cas où, je crois que ça change rien à la fonctionnalité, mais tu peux écrire la même chose en bénéficiant des syntaxes kotlin: 
`Firebase.auth.currentUser`

**Review by Huntsekkern — APPROVED** [review]
Already approving because it's close enough of what we want, and I'll already merge your branch in mine to reduce future merge issues.

Yeah, I changed a bit the interface, so it will mess up a couple of things, but so I can already take care of it.

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 2 above)
Comment (Huntsekkern) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:18]: J'ai mergé sa branche, donc je propose que j'implémente tout de suite ça dans la mienne (sinon je vais devoir rechanger mon code plus tard), modifie cette ligne et qu'on laisse passer ça ici.

Comment (Huntsekkern) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:18]: Question de plus, peut-être pour @DGoedtkindtEPFL cette fois,
https://firebase.google.com/docs/emulator-suite/connect_auth utilise 9099 comme port
alors que https://firebase.google.com/docs/emulator-suite/connect_firestore utilise 8080 comme port comme toi tu l'as fait.

Tu sais si je devrais unifier l'auth pour aussi être sur 8080 ou si c'est justement nécessaire de les avoir sur des ports différents?

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 4 above)
Comment (Huntsekkern) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:22]: Bon, j'imagine que ça complique pas mal le code, car il faudrait que la listActivity puisse avoir des subsections, pas sûr qu'elle a été pensée pour... A toi de voir si c'est facile à implémenter et si ça vaut la peine

**Comment by Huntsekkern** [comment]
Un truc de plus:
Est-ce qu'un not logged user peut accéder à "edit my interests" ? Il me semble pas, le bouton est gatekeeped dans le user profile...

ça serait cool, d'implémenter ça, pas forcément dans cette PR, sinon tout ce que je code pour un unlogged user sert à rien.

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 4 above)
Comment (Huntsekkern) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:22]: Mais tiens-moi au courant, car si on se sert jamais de savoir quels intérêts sont un field, semester ou courses, je peux aussi simplifier ma partie en giclant le triple comme Nicolas l'avait suggéré dans une PR précédente.

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 2 above)
Comment (DGoedtkindtEPFL) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:18]: Je ne sais pas, mais je n'y toucherais pas. ça pose problème?

File: app/src/main/java/com/github/polybooks/utils/UserUtil.kt [hunk 5]
```diff
@@ -0,0 +1,10 @@
+package com.github.polybooks.utils
+
+import com.github.polybooks.core.LocalUser
+import com.github.polybooks.core.LoggedUser
+import com.github.polybooks.core.User
+import com.google.firebase.auth.FirebaseUser
+
+fun fireBaseUsertoUser(firebaseUser: FirebaseUser?): User {
+    return firebaseUser?.let {LoggedUser(firebaseUser.uid, firebaseUser.displayName)}?:LocalUser
+}
```
Comment (RaulinN) [inline app/src/main/java/com/github/polybooks/utils/UserUtil.kt:10]:
```suggestion
fun firebaseUserToUser(firebaseUser: FirebaseUser?): User {
    return firebaseUser?.let {LoggedUser(firebaseUser.uid, firebaseUser.displayName)} ?: LocalUser
}
```

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 4 above)
Comment (RaulinN) [inline app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt:22]: Mhmmm... Ce serait pas un peu casser le principe des CompletableFutures ca? 🤔
Je pense que ces variables devraient être des CompletableFutrue<Set<...>> et CompletableFuture<List<Interest>>

**Review by RaulinN — APPROVED** [review]

File: app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt (hunk 4 above)
Comment (Snowflake8) [reply app/src/main/java/com/github/polybooks/adapter/database/InterestAdapter.kt]: Fait

**Pull request merged**